File: backend/app/repositories/EnrollmentRepository.scala
package repositories

import com.datastax.oss.driver.api.core.cql.Row
import models._
import services.CassandraService
import javax.inject._
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters._
import java.time.Instant
import java.util.UUID
import play.api.Logger

@Singleton
class EnrollmentRepository @Inject()(
  cassandra: CassandraService
)(implicit ec: ExecutionContext) {
  
  private val logger = Logger(this.getClass)
  private val keyspace = "subjectplanning"
  
  def findByStudentId(studentId: UUID): Future[List[Enrollment]] = {
    val query = s"""
      SELECT * FROM $keyspace.enrollments 
      WHERE student_id = $studentId 
      ALLOW FILTERING
    """
    
    cassandra.executeAsync(query).map { resultSet =>
      resultSet.currentPage().asScala.map(mapRowToEnrollment).toList
    }.recover {
      case e: Exception =>
        logger.error(s"Error finding enrollments: ${e.getMessage}", e)
        List.empty
    }
  }
  
  def findByStudentAndSemester(studentId: UUID, semester: Int): Future[List[Enrollment]] = {
    val query = s"""
      SELECT * FROM $keyspace.enrollments 
      WHERE student_id = $studentId AND semester = $semester 
      ALLOW FILTERING
    """
    
    cassandra.executeAsync(query).map { resultSet =>
      resultSet.currentPage().asScala.map(mapRowToEnrollment).toList
    }.recover {
      case e: Exception =>
        logger.error(s"Error finding enrollments by semester: ${e.getMessage}", e)
        List.empty
    }
  }
  
  def findById(id: UUID): Future[Option[Enrollment]] = {
    val query = s"""
      SELECT * FROM $keyspace.enrollments 
      WHERE id = $id
    """
    
    cassandra.executeAsync(query).map { resultSet =>
      val row = resultSet.one()
      if (row != null) Some(mapRowToEnrollment(row))
      else None
    }.recover {
      case e: Exception =>
        logger.error(s"Error finding enrollment: ${e.getMessage}", e)
        None
    }
  }
  
  def create(enrollment: Enrollment): Future[Enrollment] = {
    val completedAtStr = enrollment.completedAt.map(t => s"'$t'").getOrElse("null")
    val gradeStr = enrollment.grade.map(g => s"'$g'").getOrElse("null")
    
    val query = s"""
      INSERT INTO $keyspace.enrollments 
      (id, student_id, course_id, semester, grade, status, attendance_rate, enrolled_at, completed_at)
      VALUES (
        ${enrollment.id},
        ${enrollment.studentId},
        ${enrollment.courseId},
        ${enrollment.semester},
        $gradeStr,
        '${enrollment.status}',
        ${enrollment.attendanceRate},
        '${enrollment.enrolledAt}',
        $completedAtStr
      )
    """
    
    cassandra.executeAsync(query).map { _ =>
      logger.info(s"Created enrollment: ${enrollment.id}")
      enrollment
    }.recover {
      case e: Exception =>
        logger.error(s"Error creating enrollment: ${e.getMessage}", e)
        throw e
    }
  }
  
  def update(enrollment: Enrollment): Future[Enrollment] = {
    val gradeStr = enrollment.grade.map(g => s"'$g'").getOrElse("null")
    val completedAtStr = enrollment.completedAt.map(t => s"'$t'").getOrElse("null")
    
    val query = s"""
      UPDATE $keyspace.enrollments 
      SET grade = $gradeStr,
          status = '${enrollment.status}',
          attendance_rate = ${enrollment.attendanceRate},
          completed_at = $completedAtStr
      WHERE id = ${enrollment.id}
    """
    
    cassandra.executeAsync(query).map { _ =>
      logger.info(s"Updated enrollment: ${enrollment.id}")
      enrollment
    }.recover {
      case e: Exception =>
        logger.error(s"Error updating enrollment: ${e.getMessage}", e)
        throw e
    }
  }
  
  def getCompletedCount(studentId: UUID): Future[Int] = {
    findByStudentId(studentId).map { enrollments =>
      enrollments.count(e => e.status == CourseStatus.COMPLETED)
    }
  }
  
  def getTotalCredits(studentId: UUID): Future[Int] = {
    // This would need to join with courses - simplified here
    findByStudentId(studentId).map { enrollments =>
      enrollments.count(_.status == CourseStatus.COMPLETED) * 3 // Assuming 3 credits per course
    }
  }
  
  def getAverageAttendance(studentId: UUID): Future[Double] = {
    findByStudentId(studentId).map { enrollments =>
      if (enrollments.isEmpty) 0.0
      else enrollments.map(_.attendanceRate).sum / enrollments.length
    }
  }
  
  private def mapRowToEnrollment(row: Row): Enrollment = {
    Enrollment(
      id = row.getUuid("id"),
      studentId = row.getUuid("student_id"),
      courseId = row.getUuid("course_id"),
      semester = row.getInt("semester"),
      grade = Option(row.getString("grade")),
      status = CourseStatus.withName(row.getString("status")),
      attendanceRate = row.getDouble("attendance_rate"),
      enrolledAt = row.getInstant("enrolled_at"),
      completedAt = Option(row.getInstant("completed_at"))
    )
  }
}

File: backend-python/app/repositories/enrollment_repository.py
from typing import Optional, List
from uuid import UUID
import logging
from app.models import Enrollment, CourseStatus
from app.services.cassandra_service import cassandra_service

logger = logging.getLogger(__name__)


class EnrollmentRepository:
    def __init__(self):
        self.session = cassandra_service.get_session()
        self.keyspace = "subjectplanning"
    
    def find_by_id(self, enrollment_id: UUID) -> Optional[Enrollment]:
        """Find enrollment by ID"""
        try:
            query = f"""
                SELECT * FROM {self.keyspace}.enrollments 
                WHERE id = %s
            """
            result = self.session.execute(query, (enrollment_id,))
            row = result.one()
            
            if row:
                return self._map_row_to_enrollment(row)
            return None
        except Exception as e:
            logger.error(f"Error finding enrollment by id: {str(e)}")
            return None
    
    def find_by_student_id(self, student_id: UUID) -> List[Enrollment]:
        """Find all enrollments for a student"""
        try:
            query = f"""
                SELECT * FROM {self.keyspace}.enrollments 
                WHERE student_id = %s 
                ALLOW FILTERING
            """
            result = self.session.execute(query, (student_id,))
            return [self._map_row_to_enrollment(row) for row in result]
        except Exception as e:
            logger.error(f"Error finding enrollments by student_id: {str(e)}")
            return []
    
    def find_by_course_id(self, course_id: UUID) -> List[Enrollment]:
        """Find all enrollments for a course"""
        try:
            query = f"""
                SELECT * FROM {self.keyspace}.enrollments 
                WHERE course_id = %s 
                ALLOW FILTERING
            """
            result = self.session.execute(query, (course_id,))
            return [self._map_row_to_enrollment(row) for row in result]
        except Exception as e:
            logger.error(f"Error finding enrollments by course_id: {str(e)}")
            return []
    
    def find_current_enrollments(self, student_id: UUID) -> List[Enrollment]:
        """Find current (enrolled) courses for a student"""
        all_enrollments = self.find_by_student_id(student_id)
        return [e for e in all_enrollments if e.status == CourseStatus.ENROLLED]
    
    def find_completed_enrollments(self, student_id: UUID) -> List[Enrollment]:
        """Find completed courses for a student"""
        all_enrollments = self.find_by_student_id(student_id)
        return [e for e in all_enrollments if e.status == CourseStatus.COMPLETED]
    
    def create(self, enrollment: Enrollment) -> Enrollment:
        """Create a new enrollment"""
        try:
            query = f"""
                INSERT INTO {self.keyspace}.enrollments 
                (id, student_id, course_id, semester, grade, status, attendance_rate, enrolled_at, completed_at)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
            """
            
            self.session.execute(query, (
                enrollment.id,
                enrollment.student_id,
                enrollment.course_id,
                enrollment.semester,
                enrollment.grade,
                enrollment.status.value,
                enrollment.attendance_rate,
                enrollment.enrolled_at,
                enrollment.completed_at
            ))
            
            logger.info(f"Created enrollment: {enrollment.id}")
            return enrollment
        except Exception as e:
            logger.error(f"Error creating enrollment: {str(e)}")
            raise
    
    def update(self, enrollment: Enrollment) -> Enrollment:
        """Update an existing enrollment"""
        try:
            query = f"""
                UPDATE {self.keyspace}.enrollments 
                SET grade = %s, status = %s, attendance_rate = %s, completed_at = %s
                WHERE id = %s
            """
            
            self.session.execute(query, (
                enrollment.grade,
                enrollment.status.value,
                enrollment.attendance_rate,
                enrollment.completed_at,
                enrollment.id
            ))
            
            logger.info(f"Updated enrollment: {enrollment.id}")
            return enrollment
        except Exception as e:
            logger.error(f"Error updating enrollment: {str(e)}")
            raise
    
    def delete(self, enrollment_id: UUID) -> bool:
        """Delete an enrollment"""
        try:
            query = f"""
                DELETE FROM {self.keyspace}.enrollments 
                WHERE id = %s
            """
            self.session.execute(query, (enrollment_id,))
            logger.info(f"Deleted enrollment: {enrollment_id}")
            return True
        except Exception as e:
            logger.error(f"Error deleting enrollment: {str(e)}")
            return False
    
    def _map_row_to_enrollment(self, row) -> Enrollment:
        """Map Cassandra row to Enrollment model"""
        return Enrollment(
            id=row.id,
            student_id=row.student_id,
            course_id=row.course_id,
            semester=row.semester,
            grade=row.grade,
            status=CourseStatus(row.status),
            attendance_rate=row.attendance_rate,
            enrolled_at=row.enrolled_at,
            completed_at=row.completed_at
        )


# Singleton instance
enrollment_repository = EnrollmentRepository()
